rust_src/src/casetab.rs

```diff
@@ -3,100 +3,126 @@
+            }
+
+            if eqv.is_nil() {
+                eqv = make_char_table(Qcase_table.into(), Qnil).into();
+                map_char_table(Some(set_identity), Qnil, canon, eqv);
+                map_char_table(Some(shuffle), Qnil, canon, eqv);
+                self.set_extras(2, eqv);
+            }
+
+            // This is so set_image_of_range_1 can find the EQV table.
+            LispCaseTable::from(canon).set_extras(2, eqv);
+        }
     }
 
     pub fn get(&self, idx: isize) -> LispObject {
         self.0.get(idx)
     }
-}
 
-impl From<LispObject> for LispCaseTable {
-    fn from(obj: LispObject) -> Self {
-        let case_table: Option<LispCharTableRef> = obj.into();
-        if !case_table_p(case_table) {
-            wrong_type!(Qcase_table_p, obj);
-        }
-        Self(case_table.unwrap())
+    pub fn extras(&self) -> (LispObject, LispObject, LispObject) {
+        let extras = unsafe { self.0.extras.as_slice(3) };
+        (extras[0], extras[1], extras[2])
     }
-}
-
-static mut Vascii_downcase_table: LispObject = Qnil;
-static mut Vascii_upcase_table: LispObject = Qnil;
-static mut Vascii_canon_table: LispObject = Qnil;
-static mut Vascii_eqv_table: LispObject = Qnil;
 
-fn set_case_table(table: LispObject, standard: bool) -> LispObject {
-    let case_table: LispCaseTable = table.into();
-    let (mut up, mut canon, mut eqv) = case_table.extras();
+    pub fn set_extras(&mut self, idx: isize, value: LispObject) {
+        assert!(0 <= idx && idx < 3);
+        let extras = unsafe { self.0.extras.as_mut_slice(3) };
+        extras[idx as usize] = value;
+    }
+}
 
-    unsafe {
-        if up.is_nil() {
-            up = Fmake_char_table(Qcase_table, Qnil);
-            map_char_table(Some(set_identity), Qnil, table, up);
-            map_char_table(Some(shuffle), Qnil, table, up);
-            set_char_table_extras(table, 0, up);
-        }
+impl From<LispObject> for Option<LispCaseTable> {
+    fn from(obj: LispObject) -> Self {
+        let table = match Option::<LispCharTableRef>::from(obj) {
+            None => {
+                return None;
+            }
+            Some(t) => {
+                if t.purpose.eq(Qcase_table) {
+                    LispCaseTable(t)
+                } else {
+                    return None;
+                }
+            }
+        };
 
-        if canon.is_nil() {
-            canon = Fmake_char_table(Qcase_table, Qnil);
-            set_char_table_extras(table, 1, canon);
-            map_char_table(Some(set_canon), Qnil, table, table);
-        }
+        let (up, canon, eqv) = table.extras();
 
-        if eqv.is_nil() {
-            eqv = Fmake_char_table(Qcase_table, Qnil);
-            map_char_table(Some(set_identity), Qnil, canon, eqv);
-            map_char_table(Some(shuffle), Qnil, canon, eqv);
-            set_char_table_extras(table, 2, eqv);
+        if (up.is_nil() || up.is_char_table())
```

Comment:
I realize this is copied from elsewhere, but I think this is equivalent to 

```rust
up.is_maybe_char_table()
&& canon.is_maybe_char_table()
&& eqv.is_maybe_char_table()
```

where `x.is_maybe_char_table()` is equivalent to `x.is_nil() || x.is_char_table()`.

Reply:
Nice. I had not seen that style. But totally accurate.

Reply:
A recurring frustration with Emacs source is the places where nil is ok and where nil is not ok.

Reply:
Reading this again, I think the boolean is a little more complex. I believe the proper way to read this is:
```
up.is_maybe_char_table()
&& ((canon.is_nil() && eqv.is_nil())
        || (canon.is_char_table() && eqv.is_maybe_char_table()))
```

I like this idea of "is_maybe". I wonder if we can cheaply generalize it to `is_maybe!(type, value)`? Then we could have `is_maybe!(LispCons, obj)` too and likely `is_maybe!(LispString, obj)`.

Reply:
Okay, so it's not possible for canon to be a char table but eqv be nil. What a horrible expression.

Reply:
It would probably be better long-term to figure out a way to cast them as `Option<LispCharTableRef>` than come up with a new macro. I don't know exactly how that would work.

Reply:
A `Result<Option<Thing>>` would work. `Err` for completely wrong type, `None` for nil, and `Some` for the real value.

Reply:
Or perhaps we need/want a 3 way type to handle that.

Err
None
Value

But using the existing Result type means we can use `try_into`. Things to experiment with.

Completely off topic but relevant still. Places where we hits walls with using C are in part due to the use of setjmp/longjmp in the C code I think. So we step into undefined territory (literally). As we transition to pure Rust we will need to work out how to handle this. `lisp_fn` returns a Result type and having the macro generated wrapper handle it might be a way to maintain the current behavior but in a Rust manner. The `Err` value could track the depth and what not instead of only being a String.